### PR TITLE
Chainable texture filters

### PR DESCRIPTION
**Pull request by bdero**
Part of https://github.com/flutter/flutter/issues/99521.

* Add `FilterContents` class that take zero or more texture inputs and generates a result texture.
  * Any input may either be an `impeller::Texture` or another `FilterContents` which will get resolved at render time.
  * When the `FilterContents` is rendered with an entity, it draws the output texture to the parent pass via `TextureContents` and respects the entity's transform/stencil/path.
  * By default, the size of the first input texture is used as the size of the result texture, but filters may override this behavior depending on the use case. For example, a convolution filter (such as a Gaussian blur) may optionally increase the size of the result texture by a factor of the blur radius.
* Add `BlendFilterContents` filter for simple additive blending/merging textures together.
  * Currently, input textures that don't match the size of the first texture stretch.
* Add `Entity::BlendMode::kPlus` blend mode. 

Right now the test just blends 3 fixtures together. Will make fancier tests once we have less trivial filters to work with.

image

Usage:
```cpp
    auto bridge = CreateTextureForFixture("bay_bridge.jpg");
    auto boston = CreateTextureForFixture("boston.jpg");
    auto kalimba = CreateTextureForFixture("kalimba.jpg");

    // Draws kalimba and alpha blends boston over it.
    auto blend0 = FilterContents::MakeBlend(
        Entity::BlendMode::kSourceOver, {kalimba, boston});

    // Adds bridge*3 and the result of the first blend together.
    auto blend1 = FilterContents::MakeBlend(
        Entity::BlendMode::kPlus, {bridge, bridge, blend0, bridge});

    Entity entity;
    entity.SetPath(PathBuilder{}.AddRect({100, 100, 300, 300}).TakePath());
    entity.SetContents(blend1);
    entity.Render(context, pass);
```

### CODE REVIEW

**Comment by bdero** [comment]
This is ready for review

**Review by chinmaygarde — APPROVED** [review]
This is great. I left perhaps one too many nits but the mechanics look good.

File: entity/contents/filter_contents.h [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+  ~FilterContents();
+
+  /// @brief The input texture sources for this filter. All texture sources are
+  ///        expected to have or produce premultiplied alpha colors.
+  ///        Any input can either be a `Texture` or another `FilterContents`.
+  ///
+  ///        The number of required or optional textures depends on the
+  ///        particular filter's implementation.
+  void SetInputTextures(InputTextures& input_textures);
```
[the remark this answers is not in the document]
Comment (bdero) [reply entity/contents/filter_contents.h]: Whoops! Thanks for the great explanation. Done.

File: entity/contents/filter_contents.cc [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+            input_textures_[0])) {
+      auto& filter =
+          std::get<std::shared_ptr<FilterContents>>(input_textures_[0]);
+      return filter->GetOutputSize();
+    } else {
+      auto& texture = std::get<std::shared_ptr<Texture>>(input_textures_[0]);
+      return texture->GetSize();
+    }
+  }
+  return {0, 0};
```
[the remark this answers is not in the document]
Comment (bdero) [reply entity/contents/filter_contents.cc]: Done.

File: entity/contents/filter_contents.h [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+      std::variant<std::shared_ptr<Texture>, std::shared_ptr<FilterContents>>;
+  using InputTextures = std::vector<InputVariant>;
+
+  static std::shared_ptr<FilterContents> MakeBlend(
+      Entity::BlendMode blend_mode,
+      InputTextures input_textures);
+
+  FilterContents();
+
+  ~FilterContents();
```
[the remark this answers is not in the document]
Comment (bdero) [reply entity/contents/filter_contents.h]: Done.

File: entity/contents/filter_contents.cc [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+  });
+  auto vtx_buffer = vtx_builder.CreateVertexBuffer(host_buffer);
+
+  VS::FrameInfo frame_info;
+  frame_info.mvp = Matrix::MakeOrthographic(ISize(1, 1));
+  frame_info.alpha = 1;
+  auto uniform_view = host_buffer.EmplaceUniform(frame_info);
+
+  for (const auto& texture : input_textures) {
+    Command cmd;
```
[the remark this answers is not in the document]
Comment (bdero) [reply entity/contents/filter_contents.cc]: Nice, moved most of this setup out.

File: entity/contents/filter_contents.cc [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+  if (!sub_command_buffer->SubmitCommands()) {
+    return std::nullopt;
+  }
+
+  return subpass_texture;
+}
+
+ISize FilterContents::GetOutputSize() const {
+  if (!input_textures_.empty()) {
+    if (std::holds_alternative<std::shared_ptr<FilterContents>>(
```
[the remark this answers is not in the document]
Comment (bdero) [reply entity/contents/filter_contents.cc]: Nice, done.

File: entity/contents/filter_contents.cc [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+  if (!sub_command_buffer->SubmitCommands()) {
+    return std::nullopt;
+  }
+
+  return subpass_texture;
+}
+
+ISize FilterContents::GetOutputSize() const {
+  if (!input_textures_.empty()) {
```
[the remark this answers is not in the document]
Comment (bdero) [reply entity/contents/filter_contents.cc]: Agreed, done.

**Pull request merged**